### ISSUE TITLE
ArrayOutOfBoundException: -1 when searching by Hash Index

### ISSUE THREAD

**Issue by dzmitryh** [Netflix/hollow#159]
Hey guys! Would like to share with you my issue.

## Description:

When I try to use hash index and then search by it I get `ArrayOutOfBoundException`

It is based on this sources https://github.com/Netflix/hollow/blob/7324212d7a892f5a04ee4d83885683042ad5a166/hollow/src/test/java/com/netflix/hollow/core/index/HollowHashIndexTest.java#L32
index example: 
```
HollowHashIndex index = new HollowHashIndex(readStateEngine, "TypeB", "", "b1.value");
```

## Stack trace:
```
java.lang.ArrayIndexOutOfBoundsException: -1

	at com.netflix.hollow.core.memory.encoding.FixedLengthElementArray.getElementValue(FixedLengthElementArray.java:94)
	at com.netflix.hollow.core.memory.encoding.FixedLengthElementArray.getElementValue(FixedLengthElementArray.java:85)
	at com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadStateShard.readString(HollowObjectTypeReadStateShard.java:216)
	at com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadState.readString(HollowObjectTypeReadState.java:188)
	at com.netflix.hollow.core.read.HollowReadFieldUtils.fieldValueEquals(HollowReadFieldUtils.java:193)
	at com.netflix.hollow.core.index.HollowHashIndex.matchIsEqual(HollowHashIndex.java:165)
	at com.netflix.hollow.core.index.HollowHashIndex.findMatches(HollowHashIndex.java:103)
	at com.netflix.hollow.core.index.HollowHashIndexTest.testIndexingColonIncludedValuesThrowsArrayOutOfBound(HollowHashIndexTest.java:95)
```

## Unit test:
I've created unit test in latest sources to reproduce it:

```
    @Test
    public void testIndexingColonIncludedValuesThrowsArrayOutOfBound() throws IOException {
        HollowObjectMapper mapper = new HollowObjectMapper(writeStateEngine);

        final String b1ContainsColon = "one:";
        mapper.add(new TypeB(null));

        roundTripSnapshot();

        HollowHashIndex index = new HollowHashIndex(readStateEngine, "TypeB", "", "b1.value");
        index.findMatches(b1ContainsColon);
    }
```
## Observations:
* if u delete `colon` symbol from `b1ContainsColon` variable the test passes;
* if u fill `new TypeB("some-value")` with some value problem is gone as well.

Let me know if u need any other details.

**Comment by patrox** [comment]
@dkoszewnik I faced the same issue as @dzmitryh:

1.) I Create a `HashIndex`: 

```java
CustomHashIndex customHashIndex = new CustomHashIndex(hollowConsumer, true, "Title", "", "id.value", "language.value");
```

2.) Then when I try to query the index for some values, it fails:

```java
customHashIndex.findTitleMatches(titleId, language))
```

And I get an identical stacktrace as @dzmitryh :

<img width="1649" alt="screen shot 2018-03-14 at 11 00 27" src="https://user-images.githubusercontent.com/476392/37395655-09448c48-2777-11e8-80cf-ffdda402b7ad.png">

Can someone from Hollow gurus, help us to understand what's the problem ?
I'm willing to provide more information, but I would need a helping hand.

**Comment by patrox** [comment]
One more thing, when I annotated the field which is used in the index (`id`) with the `@HollowInline` and rebuild the index as:

```java
CustomHashIndex customHashIndex = new CustomHashIndex(hollowConsumer, true, "Title", "", "id", "language.value");
```

the problem doesn't appear.

Any clues ?

**Comment by aidarko** [comment]
Thanks to @dzmitryh I was able to start investigation quickly.
I was able to reproduce the issue and find other ones.

As a result I've created PR with the fix and several unit tests, covering different scenarios.

**Details**:
Problem is related to null values in indexed fields.
For null fields hashOrdinal becomes -1, which cause problems later during value retrieval.
The problem could be solved in ```com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadStateShard.java```

**Solution**: 
Simply add precheck during work with objects in ```com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadStateShard.java```

**Why did it work with "one", but not "one:"?** 
Hollow was trying to search values in different buckets.
Bucket is picked here: ```com/netflix/hollow/core/index/HollowHashIndex.java:98```.
More specifically, "one" leads to bucket 3, but "one:" leads to bucket 0.
Null goes to 0 as well.

**Why @HollowInline is workaround?**
When you try to use `@HollowInline`, then it should translate to the real value without Hollow wrappers.
`HollowObjectTypeReadState.readOrdinal` returns values `>= 0`;
